src/listparser/dates.py

```diff
@@ -171,3 +172,91 @@ def parse_rfc822(date: str) -> datetime.datetime | None:
         )
     except (ValueError, OverflowError):
         return None
+
+
+def parse_iso8601(date: str) -> datetime.datetime | None:
+    """Parse ISO 8601 dates and times.
+
+    This handles the most common ISO 8601 formats including:
+    - YYYY-MM-DDThh:mm:ss.sssZ
+    - YYYY-MM-DDThh:mm:ss.sss+HH:MM
+    - YYYY-MM-DDThh:mm:ss.sss-HH:MM
+
+    For Python 3.7+ we could use datetime.fromisoformat() for most formats,
+    but using regex gives us more flexibility with variations.
+    """
+    # Main ISO 8601 pattern with optional fractional seconds and timezone
+    iso_pattern = re.compile(
+        r"^(\d{4})-"  # Year
+        r"(\d{2})-"  # Month
+        r"(\d{2})"  # Day
+        r"[T ]"  # T or space separator
+        r"(\d{2}):"  # Hour
+        r"(\d{2}):"  # Minute
+        r"(\d{2})"  # Second
+        r"(?:\.(\d+))?"  # Optional fractional seconds
+        r"(Z|[+-]\d{2}:?\d{2})?"  # Optional timezone (Z or +/-HH:MM)
+        r"$"
+    )
```

Comment:
Compiling the regex each time the function is can be improved by compiling the regex outside the function.